Changes the behaviour of sequential percent price discounts

vallka opened the pull request:
In PS 1.6 all percent discounts were applied to ORIGINAL price. In 1.7, this behaviour was changed (probably unintentionally) - now next percent discount apllies to already dicounted by previous rule price. (Thus percent discounts are multiplied rather then added as it were per ver 1.6). This behaviour conflict with the info presented in back office and in the invoice.

This simple local change restores 1.6 behaviour

<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | see screenshots below
| Type?         | bug fix 
| Category?     | CO 
| BC breaks?    | Does it break backward compatibility? no
| Deprecations? | Does it deprecate an existing feature? no
| Fixed ticket? | (optional) If this PR fixes an [Issue](https://github.com/PrestaShop/PrestaShop/issues), please write "Fixes #[issue number]" here.
| How to test?  | Please indicate how to best verify that this PR is correct.


A cart in PS 1.6.1.4:

Total price = 60
Discounts: (20% = 12) + (15% = 9) + (5% = 3) = 24


![cart-1 6](https://user-images.githubusercontent.com/12764829/48769920-ce0c9680-ecb4-11e8-8153-6a6a9ed860a4.PNG)

PS 1.7.4.2
Total price = 60
Discounts: (20% = 12) + (15%*(60*80%) = 7.2) + (5%*(60*80%)*(60*85%) = 2.04) = 21.24

![cart-1 7](https://user-images.githubusercontent.com/12764829/48769962-e977a180-ecb4-11e8-9388-e0f0b8455d12.PNG)

PS 1.7.4.2 - incorrect sum in backoffice!
![backoffice-1 7](https://user-images.githubusercontent.com/12764829/48769987-fdbb9e80-ecb4-11e8-9c17-17d725b3a146.PNG)
PS 1.7.4.2 - incorrect sum in invoice!
![invoice-1 7](https://user-images.githubusercontent.com/12764829/48769995-02805280-ecb5-11e8-9a86-20fde085acb0.PNG)

Now the corrected version for 1.7.*:
![cart-1 7-corrected](https://user-images.githubusercontent.com/12764829/48770044-1b890380-ecb5-11e8-9736-18856e070fb7.PNG)
![backoffice-1 7-corrected](https://user-images.githubusercontent.com/12764829/48770046-1e83f400-ecb5-11e8-8f13-e46a2580e113.PNG)




<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/11444)
<!-- Reviewable:end -->

vallka committed:
Changes the behaviour of sequential percent price discouts

In PS 1.6 all percent discounts were applied to ORIGINAL price. In 1.7, this behaviour was changed (probably unintentionally) - now next percent discount apllies to already dicounted by previous rule price. (Thus percent discounts are multiplied rather then added as it were per ver 1.6). This behaviour conflict with the info presented in back office and in the invoice.

This simple llocal change restores 1.6 behaviour

eternoendless commented:
@colinegin what do you think? Maybe this behavior should be set as an option in the BO.

In any case, should this new behavior be accepted, new tests have to be created as the current tests are meant to validate the "discount on original price" behavior (cc @tomlev).

colinegin commented:
Hello @vallka , thanks for your feedback, it's a very interesting topic !

You are correct, there has been a change of behaviour starting from the 1.7.4 for percentage vouchers.
This change was made on purpose, as we consider it is more logic and safer to apply vouchers (in percentage) to the new calculated cart value. This prevents from having risks of carts equal to zero or negative values. You prefer the old system ? Could you tell us the reasons why ?

Regarding the conflict with the info presented in back office and in the invoice, you are correct, this is a bug so thanks for reporting it ! I've created and added the issue to our debug roadmap : https://github.com/PrestaShop/PrestaShop/issues/11541.

Thanks again !

vallka commented:
Hi Guys,

Thank you for you response.

Let me explain again. For me, as a (say) shop owner, it really doesn't 
matter which way the discounts apply. Any way is a right way. I only 
want the thing be consistent through all the system. How it currently 
works in PS 1.7.4.2 - it sums the discount one way in the shopping cart 
and the different way - in the invoice. Our customers know arithmetic to 
some base level and they ask - why summing up discounts do not 
correspond to a final discount? this the main point. Please change the 
way it is represented in the invoice and in the BO. As I said, it as 
doesn't matter at all - for me - which way it is calculated but it 
should be consistent.

The module I have changed to restore 1.6 behaviour 
(Core/Cart/CartRow.php 
<https://github.com/PrestaShop/PrestaShop/pull/11444/files#diff-03754c08904efd8b03ec536daa8a39fc>) 
  seems to have this algorithm from the very beginning, so I doubt it was 
introduced in 1.7.4 - I think it was already in the very first release 
of 1.7. But it really doesn't matter.

So if you are to change the invoice, it's ok with me. Please correct in 
in the BO and the invoice. I only think that it may be much more time 
consuming :)


Thanks again,
Valentine

------ Original Message ------
From: "colinegin" <notifications@github.com>
To: "PrestaShop/PrestaShop" <PrestaShop@noreply.github.com>
Cc: "Valentine" <vallka@vallka.com>; "Mention" 
<mention@noreply.github.com>
Sent: 28/11/2018 15:29:54
Subject: Re: [PrestaShop/PrestaShop] Changes the behaviour of sequential 
percent price discounts (#11444)

>Hello @vallka <https://github.com/vallka> , thanks for your feedback, 
>it's a very interesting topic !
>
>You are correct, there has been a change of behaviour starting from the 
>1.7.4 for percentage vouchers.
>This change was made on purpose, as we consider it is more logic and 
>safer to apply vouchers (in percentage) to the new calculated cart 
>value. This prevents from having risks of carts equal to zero or 
>negative values. You prefer the old system ? Could you tell us the 
>reasons why ?
>
>Regarding the conflict with the info presented in back office and in 
>the invoice, you are correct, this is a bug so thanks for reporting it 
>! I've created and added the issue to our debug roadmap : 
><https://github.com/PrestaShop/PrestaShop/issues/11541>.
>
>Thanks again !
>
>—
>You are receiving this because you were mentioned.
>Reply to this email directly, view it on GitHub 
><https://github.com/PrestaShop/PrestaShop/pull/11444#issuecomment-442488154>, 
>or mute the thread 
><https://github.com/notifications/unsubscribe-auth/AMLGnWpsRnnTl9iizt72nRJeOaXUc4mIks5uzqvygaJpZM4Yq52W>.
>

---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

colinegin commented:
Thanks for your reply, I just realized the link of the issue was not displayed. It should be ok now.
I completely agree, we should fix this issue as soon as possible. We will keep you updated as soon as it is done.

Thank you for your understanding,

Coline

mickaelandrieu commented:
Hello @colinegin,

what should we do with this contribution then?

Accept it, improve it or close it to push another update?

Mickaël

colinegin commented:
Hello @mickaelandrieu ,

We can close this PR, as I've created a [new issue](https://github.com/PrestaShop/PrestaShop/issues/11541) dedicated to the issue we discussed with @vallka. The fix for this issue is currently in progress. 

Thank you !
